### PR TITLE
Remove the 2GB limit when unparsing BLOBs

### PR DESCRIPTION
**Pull request by stevedlawrence**
Due to suspensions, it is possible that we could unparse a BLOB to a
buffered DataOutputStream, which means the entire contents of the BLOB
would be stored in memory until the suspension and its DataOutputStream
were finished. Not only is this a waste of memory since the blob is all
on disk, but can also limit BLOBs to 2GB since buffered
DataOutputStreams use a Byte array for buffering.

This resolves this issue by not immediately unparsing BLOB data, but
instead adding a new BLOB DataOutputStream which writes BLOB data to the
previous DataOutputStream once that stream becomes direct. This means we
never store the full BLOB data in memory and remove the 2GB limit.

DAFFODIL-2194

### CODE REVIEW

**Comment by mbeckerle** [comment]
I may have approved this a bit hastily. Are there tests of this, showing a 2GB+ Blob going through? I didn't see them in the review.

**Comment by stevedlawrence** [comment]
I've performed manual testing. I knew that @jadams-tresys was working on fixing a related issue for parse and would adding some automted tests, so didn't want to duplicate effort. But I did want some eyes on this sooner rather than later. Josh just made PR#281, so I can modify this PR to take advantage of those tests once that's merged in.

File: daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala [hunk 1]
```diff
@@ -734,6 +906,8 @@ class BitOrderChangeException(directDOS: DirectOrBufferedDataOutputStream, finfo
   }
 }
 
+class BlobIOException(message: String) extends Exception(message) with ThinThrowable
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala]: Agree.

**Comment by stevedlawrence** [comment]
WIth PR #281 merged, I've rebased onto master and updated the bLob CLI tests to unparse after parsing. The tests need to be manual enabled so aren't run as part of CI, but I have manually verified that they pass.
